File: src/main/java/br/com/tech/challenge/service/VoteService.java
package br.com.tech.challenge.service;


import br.com.tech.challenge.domain.VoteDetail;
import br.com.tech.challenge.model.Associate;
import br.com.tech.challenge.model.Vote;
import br.com.tech.challenge.repository.VoteRepository;
import br.com.tech.challenge.service.AssociateService;
import br.com.tech.challenge.service.validation.vote.VoteValidation;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.val;
import org.springframework.stereotype.Service;

import java.util.List;


@Service
@AllArgsConstructor
public class VoteService {
    private final VoteRepository voteRepository;
    private final AssociateService associateService;
    private final List<VoteValidation> voteValidation;


    public void vote(VoteDetail voteDetail) {
        Associate associate = associateService.findAssociateByCpfOrCreateNewAssociate(voteDetail.associateDetail().cpf());
        Vote vote = new Vote(
                voteDetail.session(),
                associate,
                voteDetail.vote()
        );
        for (VoteValidation validation : voteValidation) {
            validation.validate(voteDetail);
        }
        voteRepository.save(vote);
    }
}

File: src/main/java/br/com/tech/challenge/service/VoteService.kt
package br.com.ntconsult.challenge.service

import br.com.ntconsult.challenge.domain.VoteDetail
import br.com.ntconsult.challenge.model.Vote
import br.com.ntconsult.challenge.repository.VoteRepository
import br.com.ntconsult.challenge.service.validation.vote.VoteValidation
import org.springframework.stereotype.Service

@Service
class VoteService(private val voteRepository: VoteRepository,
                  private val associateService: AssociateService,
                  private val voteValidation: List<VoteValidation>) {

    fun vote(voteDetail: VoteDetail) {
        with(voteDetail) {
            val associate = associateService.findAssociateByCpfOrCreateNewAssociate(associateDetail.cpf)
            val vote = Vote(
                    session = session,
                    associate = associate,
                    vote = vote
            )
            voteValidation.forEach {
                it.valid(voteDetail)
            }
            voteRepository.save(vote)
        }
    }

}